kythe/cxx/tools/proto_metadata_plugin.cc

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2019 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <string>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/node_hash_map.h"
+#include "absl/memory/memory.h"
+#include "absl/strings/escaping.h"
+#include "absl/strings/match.h"
+#include "absl/strings/string_view.h"
+#include "absl/strings/strip.h"
+#include "glog/logging.h"
+#include "google/protobuf/compiler/code_generator.h"
+#include "google/protobuf/compiler/cpp/cpp_generator.h"
+#include "google/protobuf/compiler/plugin.h"
+#include "google/protobuf/compiler/plugin.pb.h"
+#include "google/protobuf/io/printer.h"
+#include "google/protobuf/io/zero_copy_stream_impl.h"
+
+namespace kythe {
+namespace {
+using ::google::protobuf::FileDescriptor;
+using ::google::protobuf::compiler::GeneratorContext;
+using ::google::protobuf::compiler::cpp::CppGenerator;
+using ::google::protobuf::io::Printer;
+using ::google::protobuf::io::StringOutputStream;
+using ::google::protobuf::io::ZeroCopyOutputStream;
+
+constexpr absl::string_view kMetadataFileSuffix = ".meta";
+constexpr absl::string_view kHeaderFileSuffix = ".h";
+
+std::string WriteMetadata(absl::string_view metafile,
+                          absl::string_view contents,
+                          ZeroCopyOutputStream* stream) {
+  if (!absl::EndsWith(metafile, kMetadataFileSuffix)) {
+    return "Not a metadata file";
+  }
+
+  Printer printer(stream, '\n');
+  printer.PrintRaw("/* ");
+  printer.WriteRaw(metafile.data(), metafile.size());
+  printer.PrintRaw("\n");
+  std::string metadata;
+  absl::Base64Escape(contents, &metadata);
+  printer.WriteRaw(metadata.data(), metadata.size());
```

Comment:
Some tools behave badly in the presence of very long lines. It might be worth adding line breaks to keep this to 80 columns.

Reply:
A very good point, thanks. Done.